src/parser/symtab.ts

```diff
@@ -0,0 +1,1738 @@
+
+        Note that a single name can be bound to multiple objects.  If
+        is_namespace() is true, the name may also be bound to other
+        objects, like an int or list, that does not introduce a new
+        namespace.
+        */
+
+        return !!this.__namespaces;
+    }
+
+    get_namespaces() {
+        /* Return a list of namespaces bound to this name */
+        return this.__namespaces;
+    }
+
+    get_namespace() {
+        /*
+        Returns the single namespace bound to this name.
+
+        Raises ValueError if the name is bound to multiple namespaces.
+        */
+
+        if (this.__namespaces?.length != 1) {
+            throw Error("name is bound to multiple namespaces");
+            // todo: This should be a value error maybe
+        }
+
+        return this.__namespaces[0];
+    }
+}
+
+var astScopeCounter = 0;
+
+class SymbolTableScope {
+    name: string;
+    varnames: string[];
+    children: SymbolTableScope[];
+    blockType: BlockTypes;
+    isNested = false;
+    hasFree: boolean;
+    childHasFree: boolean;
+    generator: boolean;
+    varargs: boolean;
+    varkeywords: boolean;
+    returnsValue: boolean;
+    lineno: number;
+    colOffset: number;
+    endLineno: number | null | undefined;
+    endColOffset: number | null | undefined;
+    table: SymbolTable;
+    symbols: { [name: string]: number };
+    _funcLocals: string[] | null = null;
+    _funcParams: string[] | null = null;
+    _funcGlobals: string[] | null = null;
+    _funcFrees: string[] | null = null;
+    _classMethods: string[] | null = null;
+    comp_iter_expr = 0;
+    comp_iter_target = false;
+    comprehension = false;
+    directives: [string, number, number][] | null = null;
+    coroutine = false;
+
+    constructor(
+        table: SymbolTable,
+        name: string,
+        type: BlockTypes,
+        ast: AST,
+        lineno: number,
+        colOffset: number,
+        endLineno?: number | null,
+        endColOffset?: number | null
+    ) {
+        this.name = name;
+        this.varnames = [];
+        this.children = [];
+        this.blockType = type;
+
+        this.isNested = false;
+        this.hasFree = false;
+        this.childHasFree = false; // true if child block has free vars including free refs to globals
+        this.generator = false;
+        this.varargs = false;
+        this.varkeywords = false;
+        this.returnsValue = false;
+
+        this.lineno = lineno;
+        this.colOffset = colOffset;
+        this.endLineno = endLineno;
+        this.endColOffset = endColOffset;
+
+        this.table = table;
+
+        if (table.cur && (table.cur.isNested || table.cur.blockType === FunctionBlock)) {
+            this.isNested = true;
+        }
+
+        ast.scopeId = astScopeCounter++;
+        table.stss[ast.scopeId] = this;
+
+        // cache of Symbols for returning to other parts of code
+        this.symbols = {};
+    }
+
+    get_type() {
+        return this.blockType;
+    }
```

Comment:
This seems based on the python version.
And so far as I can tell we're only using it for the dump function/assert statements?
So should we include them?

Reply:
yes, I removed them in at one point thinking they should be in separate types but then added them back in later 🤷. What do you think?

Reply:
The dump function will be internal. 
And if we get round to exposing symtable in python (probably never) then we'll have to make these python methods anyway.
So i'd er on the side of we don't need them.

Reply:
actually maybe keep them for now. They might be useful and will make the dump function easier. And some of them are quite informative